### ISSUE TITLE
How to convert Matft array to OpenCV mat array

### ISSUE THREAD

**Issue by Eumentis-Jayashree** [jjjkkkjjj/Matft#53]
I have created Matft array from float array `let matftArray = MfArray(floatArrayValues)` now I want to convert this array to OpenCV mat array so that I can use this array for processing.

**Comment by jjjkkkjjj** [comment]
@Eumentis-Jayashree 
According to [OpenCV's code](https://github.com/opencv/opencv/blob/4.x/modules/imgcodecs/src/apple_conversions.mm), how about like this? (The below is psuedo code.)

```swift
void  MfarrayToMat(const MfArray array, cv::Mat& m) {
    // you should assert array is row contiguous
    Int cols = array.shape[0], rows = array.shape[1], channel = array.shape[2];

    if (channel == 1)
    {
        m.create(rows, cols, CV_32FC1);
        memcpy(array.mfdata.data_real, m.data, array.storedSize);
    }
    else if (channel == 3)
    {
        m.create(rows, cols, CV_32FC4); 
        memcpy(array.mfdata.data_real, m.data, array.storedSize);
    }
    else
    {
        m.create(rows, cols, CV_32FC4);
        memcpy(array.mfdata.data_real, m.data, array.storedSize);
    }
}
```

**Comment by Eumentis-Jayashree** [comment]
Hey @jjjkkkjjj thank you for the response.

Actually, I am getting the result by the ONNX model. So the output I get is in **ORTValue** form, which has this shape [1, 359, 8400].

1. I converted that bytedata output to the float array 
`let floatValuesForOutput = rawOutputData.withUnsafeBytes { (buffer: UnsafeRawBufferPointer) -> [Float] in
                let floatBuffer = buffer.bindMemory(to: Float.self)
                let count = rawOutputData.count / MemoryLayout<Float>.stride
                return Array(UnsafeBufferPointer(start: floatBuffer.baseAddress, count: count))
            }
`

2. After that I have converted floatValues to Matft array 
` let arr = MfArray(floatValuesForOutput)`

So, according to your solution, if I print the shape of the **arr**, then I only get Shape: [3015600], which is equal to 359 * 8400.

**Comment by jjjkkkjjj** [comment]
@Eumentis-Jayashree 

The reason of  the shape: `[3015600]` is you pass the "1d" array to MfArray.
Namely, `floatValuesForOutput ` is 1d array.

So, when you want MfArray with shape: [1, 359, 8400], you should reshape it by;

```swift
let arr = MfArray(floatValuesForOutput, shape: [1, -1, 8400])
```

Caution:
The conversion method you mentioned is asserted that `rawOutputData` is row contiguous. If you pass the NOT row contiguous `rawOutputData`, the MfArray will be different from it.

To prevent this, you should manage not only data but also shape and strides. Does `ORTValue` have strides property?

**Comment by jjjkkkjjj** [comment]
Oh, I'm sorry, the above code is invalid due to not supporting negative dimension in init...
Instead of init, you can do it like this;

```swift
let arr = MfArray(floatValuesForOutput).reshape([1, -1, 8400])
```

**Comment by Eumentis-Jayashree** [comment]
> @Eumentis-Jayashree
> 
> The reason of the shape: `[3015600]` is you pass the "1d" array to MfArray. Namely, `floatValuesForOutput ` is 1d array.
> 
> So, when you want MfArray with shape: [1, 359, 8400], you should reshape it by;
> 
> ```swift
> let arr = MfArray(floatValuesForOutput, shape: [1, -1, 8400])
> ```
> 
> Caution: The conversion method you mentioned is asserted that `rawOutputData` is row contiguous. If you pass the NOT row contiguous `rawOutputData`, the MfArray will be different from it.
> 
> To prevent this, you should manage not only data but also shape and strides. Does `ORTValue` have strides property?

No, ORTValue doesn't have strides property. It only has tensorData(), typeInfo(), tensorTypeAndShapeInfo() and tensorStringData() properties.

**Comment by jjjkkkjjj** [comment]
@Eumentis-Jayashree 
Did you solve it?
And also, I'm interested in the actual code converting MfArray into cv::Mat. If possible, could you share it to me? (I'd like to integrate it into main branch)

**Comment by Eumentis-Jayashree** [comment]
> ```swift
> shape: [1, -1, 8400]
> ```

Yeah, I solved it. But I am not sure if it's right approach. I have converted the Matft array to a Swift array and then converted it into an OpenCV mat array.

**Comment by Eumentis-Jayashree** [comment]
> @Eumentis-Jayashree Did you solve it? And also, I'm interested in the actual code converting MfArray into cv::Mat. If possible, could you share it to me? (I'd like to integrate it into main branch)

Yeah, I solved it. But I am not sure if it's right approach. I have converted the Matft array to a Swift array and then converted it into an OpenCV mat array.

**Comment by Eumentis-Jayashree** [comment]
> Hey @jjjkkkjjj thank you for the response.
> 
> Actually, I am getting the result by the ONNX model. So the output I get is in **ORTValue** form, which has this shape [1, 359, 8400].
> 
> 1. I converted that bytedata output to the float array
>    `let floatValuesForOutput = rawOutputData.withUnsafeBytes { (buffer: UnsafeRawBufferPointer) -> [Float] in let floatBuffer = buffer.bindMemory(to: Float.self) let count = rawOutputData.count / MemoryLayout<Float>.stride return Array(UnsafeBufferPointer(start: floatBuffer.baseAddress, count: count)) } `
> 2. After that I have converted floatValues to Matft array
>    ` let arr = MfArray(floatValuesForOutput)`
> 
> So, according to your solution, if I print the shape of the **arr**, then I only get Shape: [3015600], which is equal to 359 * 8400.

Here I get **rawOutputData** from ORTValue using `.tensorData()` property which data type is **`Data`**. Can I directly convert this **rawOutputData** into Matft array? so that I don't need above conversion.

**Comment by Eumentis-Jayashree** [comment]
@jjjkkkjjj The output which I get is in shape of [1, 359, 8400]. Here 359 are rows and 8400 are columns. There is no channel like image.

**Comment by jjjkkkjjj** [comment]
> Yeah, I solved it. But I am not sure if it's right approach. I have converted the Matft array to a Swift array and then converted it into an OpenCV mat array.

OK, but it is not efficient due to coping the data twice as you mentioned.

>  get rawOutputData from ORTValue using .tensorData() property which data type is Data. Can I directly convert this rawOutputData into Matft array?

That's why we should use same data or copy it once. Fortunately, I implemented borrowing data system by [this](https://github.com/jjjkkkjjj/Matft/blob/397fa338166cdf965635f302a0e48c29b741bf9f/Sources/Matft/core/object/mfdata.swift#L90);

```swift
let shape = // get from ORTValue
let size = // calculate the size from product all dimension = 1*359*8400
let floatValuesForOutput = rawOutputData.withUnsafeBytes { 
    (buffer: UnsafeRawBufferPointer) -> MfArray in 
    // copy data automatically
    let mfdata = MfData(source: nil, data_real_ptr: buffer.baseAddress!, data_imag_ptr: nil, storedSize: size, mftype: .Float, offset: 0)
    let mfstructure = MfStructure(shape: shape)
    return MfArray(mfdata: mfdata, mfstructure: mfstructure)
}
```

If the `ORTValue` conform to `MfDataBasable`, you can share data without copying, so that it is very efficient!

**Comment by Eumentis-Jayashree** [comment]
> > Yeah, I solved it. But I am not sure if it's right approach. I have converted the Matft array to a Swift array and then converted it into an OpenCV mat array.
> 
> OK, but it is not efficient due to coping the data twice as you mentioned.
> 
> > get rawOutputData from ORTValue using .tensorData() property which data type is Data. Can I directly convert this rawOutputData into Matft array?
> 
> That's why we should use same data or copy it once. Fortunately, I implemented borrowing data system by [this](https://github.com/jjjkkkjjj/Matft/blob/397fa338166cdf965635f302a0e48c29b741bf9f/Sources/Matft/core/object/mfdata.swift#L90);
> 
> ```swift
> let shape = // get from ORTValue
> let size = // calculate the size from product all dimension = 1*359*8400
> let floatValuesForOutput = rawOutputData.withUnsafeBytes { 
>     (buffer: UnsafeRawBufferPointer) -> MfArray in 
>     // copy data automatically
>     let mfdata = MfData(source: nil, data_real_ptr: buffer.baseAddress!, data_imag_ptr: nil, storedSize: size, mftype: .Float, offset: 0)
>     let mfstructure = MfStructure(shape: shape)
>     return MfArray(mfdata: mfdata, mfstructure: mfstructure)
> }
> ```
> 
> If the `ORTValue` conform to `MfDataBasable`, you can share data without copying, so that it is very efficient!

@jjjkkkjjj Thank you :). But the [this](https://github.com/jjjkkkjjj/Matft/blob/397fa338166cdf965635f302a0e48c29b741bf9f/Sources/Matft/core/object/mfdata.swift#L90) constructor is define as **internal**, I am unable to access it.

**Comment by jjjkkkjjj** [comment]
OMG!
I'll change it into public ASAP.
Wait for a moment...

**Comment by jjjkkkjjj** [comment]
@Eumentis-Jayashree 
I've changed it! Please check.

**Comment by jjjkkkjjj** [comment]
This one is no-copy (share version)

```swift
extension ORTValue: MfDataBasable{
}

let rawOutputData // ORTValue
let shape = // get from ORTValue
let size = // calculate the size from product all dimension = 1*359*8400
let floatValuesForOutput = rawOutputData.withUnsafeBytes { 
    (buffer: UnsafeRawBufferPointer) -> MfArray in 
    // share data
    let mfdata = MfData(source: rawOutputData, data_real_ptr: buffer.baseAddress!, data_imag_ptr: nil, storedSize: size, mftype: .Float, offset: 0)
    let mfstructure = MfStructure(shape: shape)
    return MfArray(mfdata: mfdata, mfstructure: mfstructure)
}
```

**Comment by Eumentis-Jayashree** [comment]
> ```swift
> let floatValuesForOutput = rawOutputData.withUnsafeBytes { 
>     (buffer: UnsafeRawBufferPointer) -> MfArray in 
>     // copy data automatically
>     let mfdata = MfData(source: nil, data_real_ptr: buffer.baseAddress!, data_imag_ptr: nil, storedSize: size, mftype: .Float, offset: 0)
>     let mfstructure = MfStructure(shape: shape)
>     return MfArray(mfdata: mfdata, mfstructure: mfstructure)
> }
> ```



> > > Yeah, I solved it. But I am not sure if it's right approach. I have converted the Matft array to a Swift array and then converted it into an OpenCV mat array.
> > 
> > 
> > OK, but it is not efficient due to coping the data twice as you mentioned.
> > > get rawOutputData from ORTValue using .tensorData() property which data type is Data. Can I directly convert this rawOutputData into Matft array?
> > 
> > 
> > That's why we should use same data or copy it once. Fortunately, I implemented borrowing data system by [this](https://github.com/jjjkkkjjj/Matft/blob/397fa338166cdf965635f302a0e48c29b741bf9f/Sources/Matft/core/object/mfdata.swift#L90);
> > ```swift
> > let shape = // get from ORTValue
> > let size = // calculate the size from product all dimension = 1*359*8400
> > let floatValuesForOutput = rawOutputData.withUnsafeBytes { 
> >     (buffer: UnsafeRawBufferPointer) -> MfArray in 
> >     // copy data automatically
> >     let mfdata = MfData(source: nil, data_real_ptr: buffer.baseAddress!, data_imag_ptr: nil, storedSize: size, mftype: .Float, offset: 0)
> >     let mfstructure = MfStructure(shape: shape)
> >     return MfArray(mfdata: mfdata, mfstructure: mfstructure)
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If the `ORTValue` conform to `MfDataBasable`, you can share data without copying, so that it is very efficient!
> 
> @jjjkkkjjj Thank you :). But the [this](https://github.com/jjjkkkjjj/Matft/blob/397fa338166cdf965635f302a0e48c29b741bf9f/Sources/Matft/core/object/mfdata.swift#L90) constructor is define as **internal**, I am unable to access it.

I am getting error _**Cannot convert value of type 'UnsafeRawPointer' to expected argument type 'UnsafeMutableRawPointer'**_ at _data_real_ptr: buffer.baseAddress!_ , because constructor accepts this data type **data_real_ptr: UnsafeMutableRawPointer**
Do I need to convert this (buffer: UnsafeRawBufferPointer) to UnsafeMutableRawPointer?

**Comment by jjjkkkjjj** [comment]
@Eumentis-Jayashree 
Does `rawOutputData` have `withUnsafeMutableBytes` method? If so, when you replace `withUnsafeBytes` into  `withUnsafeMutableBytes`, the error will be suppressed.
Otherwise, I guess you can convert `UnsafeRawPointer ` into `UnsafeMutableRawPointer ` by;

```swift
let mfdata = MfData(source: nil, data_real_ptr: UnsafeMutableRawPointer(mutating: buffer.baseAddress!), data_imag_ptr: nil, storedSize: size, mftype: .Float, offset: 0)
```

**Comment by Eumentis-Jayashree** [comment]
I have converted UnsafeRawBufferPointer to UnsafeMutableRawPointer it's giving multiple errors at the time of build.


Showing All Messages
../Matft/core/util/typeconversion.swift:73:19: Incorrect argument label in call (have 'from:count:', expected 'repeating:count:')

../Matft/core/util/typeconversion.swift:73:40: Cannot convert value of type 'UnsafePointer<T>' to expected argument type 'T'

../Matft/core/util/typeconversion.swift:134:24: Incorrect argument label in call (have 'from:count:', expected 'repeating:count:')

../Matft/core/util/typeconversion.swift:134:45: Cannot convert value of type 'UnsafeMutablePointer<Float>' to expected argument type 'Float'

../Matft/core/util/typeconversion.swift:144:18: Value of type 'UnsafeMutablePointer<Float>' has no member 'moveUpdate'

../Matft/core/util/typeconversion.swift:154:24: Incorrect argument label in call (have 'from:count:', expected 'repeating:count:')

../Matft/core/util/typeconversion.swift:154:72: Cannot convert value of type 'UnsafePointer<Float>' to expected argument type 'Float'

../Matft/core/util/typeconversion.swift:234:18: Value of type 'UnsafeMutablePointer<Double>' has no member 'moveUpdate'

../Matft/core/util/typeconversion.swift:241:24: Incorrect argument label in call (have 'from:count:', expected 'repeating:count:')

../Matft/core/util/typeconversion.swift:241:45: Cannot convert value of type 'UnsafeMutablePointer<Double>' to expected argument type 'Double'

../Matft/core/util/typeconversion.swift:256:24: Incorrect argument label in call (have 'from:count:', expected 'repeating:count:')

../Matft/core/util/typeconversion.swift:256:73: Cannot convert value of type 'UnsafePointer<Double>' to expected argument type 'Double'

../Matft/core/util/typeconversion.swift:269:27: Value of type 'UnsafeMutablePointer<U>' has no member 'moveUpdate'

**Comment by jjjkkkjjj** [comment]
Which swift version was used?

**Comment by Eumentis-Jayashree** [comment]
swift-driver version: 1.62.15 Apple Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51)
Target: arm64-apple-macosx13.0

**Comment by Eumentis-Jayashree** [comment]
swift 5

**Comment by jjjkkkjjj** [comment]
Hmm, It's strange.
My code can build correctly.

```swift
let d: [Float] = [2, 3] // I didn't install ORTValue. So I use a simple array as workaround
let size = 2
let shape = [2,1]
let floatValuesForOutput = d.withUnsafeBytes {
                (buffer: UnsafeRawBufferPointer) -> MfArray in
                // copy data automatically
                let mfdata = MfData(source: nil, data_real_ptr: UnsafeMutableRawPointer(mutating: buffer.baseAddress!), data_imag_ptr: nil, storedSize: size, mftype: .Float, offset: 0)
                let mfstructure = MfStructure(shape: shape, mforder: .Row)
                return MfArray(mfdata: mfdata, mfstructure: mfstructure)
            }
```

**Comment by jjjkkkjjj** [comment]
@Eumentis-Jayashree 
Which Matft version did you use? Is it latest version?

**Comment by jjjkkkjjj** [comment]
@Eumentis-Jayashree 
Namely, the codes updated after 0.3.3 caused this problem...
This [commit](https://github.com/jjjkkkjjj/Matft/commit/bf07973b3c33d91cfb5d2af2cc013b53aef53b7c) may be the reason.
I reverted it on the `revert_duplicate` branch. Could you try the codes again in `revert_duplicate` branch?

**Comment by Eumentis-Jayashree** [comment]
@jjjkkkjjj It's working on the `revert_duplicate` branch. Thank you :)

**Comment by Eumentis-Jayashree** [comment]
I have converted Matft array to openCV mat. 
**Steps:** 

1. Converted Matft array to NSData

       func matftArrayToNSData2(mfArray: MfArray) -> NSData? {
          guard mfArray.mftype == .Float else {
          print("MfArray is not of type Float. Actual type: \(mfArray.mftype)")
          return nil
       }
    
       // Flatten the MfArray if it's multidimensional
         let flatMfArray = mfArray.flatten()

       // Convert MfArray to Swift array of type [Float]
       guard let swiftArray =  flatMfArray.astype(.Float).toArray() as? [Float] else {
         print("Failed to convert MfArray to [Float]")
         return nil
       }

       // Convert the Swift array to Data
       let data = NSData(bytes: swiftArray, length: swiftArray.count * MemoryLayout<Float>.size)

        return data 
       }

2. Created function in objective c to convert NSData to mat
 ```
 +(NSArray *) matFromNSData:(NSData *)data rows:(int)rows cols:(int)cols{
    // Convert NSData to a byte array
     const unsigned char *byteArray = (const unsigned char *)[data bytes];
    
    // Create a cv::Mat object from the byte array
    cv::Mat mat(rows, cols, CV_8UC4, (void *)byteArray);
    
   // ........performed some operations & converted mat to NSArray
  // return NSArray 
   return nsArray
}
```

**Comment by jjjkkkjjj** [comment]
@Eumentis-Jayashree 
Thank you for sharing your codes!
I'll try to implement the method to convert MfArray to opencv's mat later referencing your code!
